File: Data Structures & Algorithms/linked-list-cycle-detection/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        if(!head) return false;
        ListNode* slowptr = head;
        ListNode* fastptr = head->next;
        while(fastptr && fastptr->next) {
            if(fastptr == slowptr) {
                return true;
            }
            slowptr = slowptr->next;
            fastptr = fastptr->next->next;
        }
        return false;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        qmin = []

        for i, node in enumerate(lists):
            if node:
                heapq.heappush(qmin, (node.val, i, node)) # i is there just for tie breaker as diff. bt c++ & py.
        
        res = ListNode()
        curr = res
        while qmin:
            val, i, node = heapq.heappop(qmin)

            curr.next = ListNode(val)
            curr = curr.next

            if node.next:
                heapq.heappush(qmin, (node.next.val, i, node.next))
        
        return res.next